[turnstile] advanced Cloudflare Turnstile example + improved client-side rendering docs (#7858)
cloudflare/cloudflare-docs: content/turnstile/get-started/client-side-rendering.md

`https://challenges.cloudflare.com/turnstile/v0/api.js?render=explicit`

Or:

`https://challenges.cloudflare.com/turnstile/v0/api.js?onload=onloadTurnstileCallback`

> When using `render=explicit`, HTML elements with the `cf-turnstile` class will not show a challenge. The `turnstile.render` function must be invoked using the following steps.  To combine both options, you would pass a querystring of `?render=explicit&onload=onloadTurnstileCallback`:

Reviewer: ```suggestion
When using `render=explicit`, HTML elements with the `cf-turnstile` class will not show a challenge. The `turnstile.render` function must be invoked using the following steps.  To combine both options, pass a query string of `?render=explicit&onload=onloadTurnstileCallback`:
```

> When using `render=explicit`, HTML elements with the `cf-turnstile` class will not show a challenge. The `turnstile.render` function must be invoked using the following steps.  To combine both options, you would pass a querystring of `?render=explicit&onload=onloadTurnstileCallback`:

Reviewer 2: should be fixed now 👍

`https://challenges.cloudflare.com/turnstile/v0/api.js?render=explicit&onload=onloadTurnstileCallback`

## Explicitly render the Turnstile widget

> 1. Insert the JavaScript tag and related code below (and be sure to replace `<YOUR_SITE_KEY>` with your site key).  **Also ensure that you have renamed the class name of `.cf-turnstile` to `#example-container` (if and only if you do not set the `render=explicit` querystring option as shown above, because it will _still_ render otherwise).**  Once the script is embedded, you will have access to a global function with multiple callback options you can customize. For the following function to work properly, the page must contain an HTML element with ID `example-container`.<br>The challenge can be invoked explicitly with the following JavaScript snippet:

Reviewer: ```suggestion
1. Insert the JavaScript tag and the related code below. Ensure that you have renamed the class name of `.cf-turnstile` to `#example-container` (if you do not set the `render=explicit` query string option as shown above, it will still render otherwise). Once the script is embedded, you will have access to a global function with multiple callback options you can customize. For the following function to work properly, the page must contain an HTML element with ID `example-container`.
```

        },
    });
};
```
</div>

Or:

<div>

```html
<script src="https://challenges.cloudflare.com/turnstile/v0/api.js?render=explicit"></script>
```
</div>

<div>

```javascript
// if using synchronous loading, will be called once the DOM is ready
turnstile.ready(function () {
    turnstile.render('#example-container', {
        sitekey: '<YOUR_SITE_KEY>',
        callback: function(token) {
            console.log(`Challenge Success ${token}`);
        },
    });
});
```
</div>

Turnstile can be programmatically loaded by invoking the `turnstile.render()` function in the global `window` object.

The `turnstile.render: function (container: string | HTMLElement, params: RenderParameters)` render takes an argument to a HTML widget.

If the invocation is successful, the function returns a `widgetId (string)`. If the invocation is unsuccessful, the function returns `undefined`.

Check out the [demo](https://demo.turnstile.workers.dev/explicit) and its [source code](https://github.com/cloudflare/turnstile-demo-workers/blob/main/src/explicit.html).

> Here's an advanced example for dynamically rendering Turnstile widgets using Bootstrap 3 or 4 with modals and responsive widget sizes:

Reviewer: ```suggestion
This is an advanced example for dynamically rendering Turnstile widgets using Bootstrap 3 or 4 with modals and responsive widget sizes:
```

        },
    });
};
```
</div>

Or:

<div>

```html
<script src="https://challenges.cloudflare.com/turnstile/v0/api.js?render=explicit"></script>
```
</div>

<div>

```javascript
// if using synchronous loading, will be called once the DOM is ready
turnstile.ready(function () {
    turnstile.render('#example-container', {
        sitekey: '<YOUR_SITE_KEY>',
        callback: function(token) {
            console.log(`Challenge Success ${token}`);
        },
    });
});
```
</div>

Turnstile can be programmatically loaded by invoking the `turnstile.render()` function in the global `window` object.

The `turnstile.render: function (container: string | HTMLElement, params: RenderParameters)` render takes an argument to a HTML widget.

If the invocation is successful, the function returns a `widgetId (string)`. If the invocation is unsuccessful, the function returns `undefined`.

Check out the [demo](https://demo.turnstile.workers.dev/explicit) and its [source code](https://github.com/cloudflare/turnstile-demo-workers/blob/main/src/explicit.html).

> Here's an advanced example for rendering Turnstile widgets using Bootstrap 3 or 4 with modals and responsive widget sizes:

Reviewer 3: ```suggestion
Here's an advanced example for dynamically rendering Turnstile widgets using Bootstrap 3 or 4 with modals and responsive widget sizes:
```

<div>

```javascript
//
// Handle explicit Cloudflare Turnstile (Advanced Example with Bootstrap/Modals/Responsive Support)
// <https://github.com/forwardemail/forwardemail.net>
// <https://developers.cloudflare.com/turnstile/get-started/client-side-rendering/#explicitly-render-the-turnstile-widget>
//
function handleExplicitTurnstile() {
  let widgetId = $(this).data('widgetId');
  if (widgetId) return;

  const isModal = $(this).parents('.modal:first').length > 0;

>   widgetId = window.turnstile.render(this, {

Reviewer 3: I don't see a call to `turnstile.remove`, so this will slowly leak memory over time.

    sitekey: '<YOUR_SITE_KEY>',
    tabindex: isModal ? -1 : 0,
    action: isModal ? 'modal' : 'page',

>     'expired-callback': (err) => console.error(err),
>     'timeout-callback': (err) => console.error(err),

Reviewer 3: ```suggestion
```
There is no need to handle these (they dont take error-related arguments either).

    'error-callback': (err) => console.error(err)
    // uncomment this if you want to render a smaller size (e.g. 576px = Bootstrap "md" breakpoint)
    // size: window.matchMedia('(min-width: 576px)').matches ? 'normal' : 'compact'
  });

  if (widgetId) {
    $(this).data('widgetId', widgetId);
    return;
  }

  console.error(new Error('[Cloudflare Turnstile] widgetId missing'));
  alert('Turnstile render error, please try again or contact us.');

>   window.location.reload();

Reviewer 3: This is impossible, turnstile always returns a widget id or throws an error.
